### DISCUSSION TITLE
Updating PullRequest generator apps

### DISCUSSION THREAD

**Discussion by pareeohnos** [argoproj/argo-cd#9453] [Q&A] [answered]
I'm implementing ArgoCD into our workflow to deploy review apps from pull requests, and I've got about 90% of the way there. I have a `PullRequest` generator `ApplicationSet` defined which creates a new app for any pull request labelled `deploy`. When I add the label to the pull request, I trigger a build of the container and push this to the registry. Once this is done, I can synchronise the app within Argo and everything is up and running which is great.

What I'm unsure about however, is how to fully automate this process. I want to get to a point ideally where once the pull request app is deployed, I can just push new changes to the code, and the deployment automatically updates. Obviously this is exactly what `AutoSync` is for, but my concern here is that the sync will trigger before the image has been updated.

For example, lets say I've got a PR open and already deployed on the cluster. Now I make some changes and push the updated code. ArgoCD will be notified via a web hook that there was a change, and update the application accordingly (`image.tag` helm variable based on commit sha), however the CI is still in the middle of building the image for that commit.

In this example, what exactly is going to happen? Is Argo going to try syncing, and just have the app sit there unable to pull the image because it doesn't exist? Is this the right way to go and how best to handle it, or is there a better solution I'm missing. I've seen the `image-updater` project which might do what I need but would like some input from the community

**Comment by rishabh625** [comment]
I think ImageUpdater is what you are looking for!! but 
I also think of a work around, that you could configure a Job that may have a script to look that new image is being added and once image is updated job would be considered as completed.

If this job is added as pre sync hook it will always ensure that image exists before syncing kubernetes resources via application.

**Reply by pareeohnos** [reply]
It definitely sounds like exactly what the image updater was created for I figured that would be the way to go but wanted to know if there were other options. The other workaround I'd thought of was using the ArgoCD CLI within the CI to instruct it to sync when the image has been pushed

**Comment by lukonjun** [comment]
Hi @pareeohnos. Regarding your last point `Is Argo going to try syncing, and just have the app sit there unable to pull the image because it doesn't exist?` ... 
  
If I understand your set up correctly Argo would update you Kubernetes Manifest, in this case I am assuming your Deployment, and change the image tag dynamically for every PR with the label deploy. Your Deployment (Kubernetes Object) would start a new pod that will try to pull the image with the new tag, however as long as it cant not pull the new Image and get into a healthy state, the old pod would also not be deleted, so I see not problem with your set up. 

If you want to look into a set up where you have an application code repository and an infra repository (Kubernetes Manifests) then image-updater can be interesting imo.

**Reply by pareeohnos** [reply]
Yeah that makes sense.

I've already separated the app repo and infra repo so using the image-updater would be fine.

**Comment by pareeohnos** [comment, marked as answer]
Went for different approach, and have got the CI calling the CD explicitly to sync when it's finished building. Came across [this part](https://argo-cd.readthedocs.io/en/stable/user-guide/ci_automation/) of the documentation and it felt like a better fit
